scalyr_agent/sharded_copying_manager/copying_manager.py

```diff
@@ -1204,99 +1137,94 @@ def _scan_for_new_logs_if_necessary(
             log_matchers, checkpoints=checkpoints, copy_at_index_zero=copy_at_index_zero
         )
 
+    def __find_and_read_checkpoints(self):
+        # type: () -> Dict
+        """
+        Find and read checkpoints that were previously written.
+        During its work, the worker writes its own checkpoint files. The copying manager, before the beginning,
+        reads them and combines into one checkpoints dict.
+        :return:
+        """
+
+        checkpoints_dir_path = os.path.join(
+            self.__config.agent_data_path, "checkpoints"
+        )
+        # get checkpoint file names from the 'checkpoints' folder. All previous worker checkpoints are stored there.
+        checkpoints_file_paths = list(glob.glob(os.path.join(checkpoints_dir_path, "checkpoints*.json")))
+
+        # we also add path for the checkpoint file that was used in the previous version of the Copying manager.
+        # This is needed to be able to get checkpoints after the upgrade of the agent.
+        checkpoints_file_paths.append(
+            os.path.join(self.__config.agent_data_path, "checkpoints.json")
+        )
+
+        found_checkpoints = []
+
+        current_time = time.time()
+
+        for checkpoints_path in checkpoints_file_paths:
 
-    def __get_checkpoints(self, current_time):
-        """Find and read worker checkpoints that were previously written."""
+            checkpoints = self.__read_checkpoint_state(
+                checkpoints_path, current_time
+            )
 
-        checkpoints = self.__find_and_read_workers_checkpoints(current_time)
+            if checkpoints:
+                found_checkpoints.append(checkpoints)
 
-        if not checkpoints:
+        if len(found_checkpoints) == 0:
             log.info(
                 "No checkpoints were found. All logs will be copied starting at their current end"
             )
 
-        result = {}
+            return {}
 
+        result = {}
         # merge checkpoints from  all workers to one checkpoint.
 
         # checkpoints from different workers may contain checkpoint for the same file,
         # so we sort checkpoints by time and update resulting collection with the same order.
-        checkpoints.sort(key=operator.itemgetter("time"))
+        found_checkpoints.sort(key=operator.itemgetter("time"))
 
-        for wc in checkpoints:
+        for wc in found_checkpoints:
```

Comment:
Right now, it doesn't seem to me like there's anything that will delete the old style checkpoint files.  (i.e., after upgrade, what would delete the `checkpoints.json` and `active-checkpoints.json` files?  For that matter, if the number of workers decrease, the checkpoints for the previous workers will never be deleted.

I think to do some garbage collection, we should first write out the full checkpoint state to `checkpoints.json` and then delete all the other checkpoint files in this directory (both active and regular).  That way we at least try to clean up old state.

Reply:
There is a new member function called `__prepare_checkpoints`. If creates a consolidated master checkpoint file(including reading old-style checkpoints) and, after the creation, it removes all checkpoints(except master), including old-style ones. 
https://github.com/scalyr/scalyr-agent-2/blob/137cde17d5a53d14102a58b1baaec22a9d93c0b1/scalyr_agent/sharded_copying_manager/copying_manager.py#L1216